### ISSUE TITLE
Channel Features

### ISSUE THREAD

**Issue by dignifiedquire** [stjepang/piper#1]
I was wondering if you have considered adding 

- `try_send`, given that there is already `try_recv`,
- `send_all`, or some form of broadcasting 

to the channels.

**Comment by jjl** [comment]
I was going to file an issue about try_send because I noticed sending to a closed socket just awaits `futures::future::pending`, but I was beaten to it :)

**Comment by jjl** [comment]
I was going to offer a PR for this, but when I went to make it I realised there's quite a lot of potential ways it could go, so it's probably best to discuss first.

**Comment by stjepang** [comment]
Yeah, I'm also a bit stuck on this :) There are two workarounds:

1. Do `select!` over a send operation and a 'default' case.
2. Do `sender.send(message).now_or_never()`.

**Comment by jjl** [comment]
I wonder how you'd feel about sending into a closed channel simply pretending to have succeeded?

**Comment by stjepang** [comment]
If the channel is closed, i think it'd be best to do put some kind of pressure onto the sending task. Some options:

1. Make the sending task panic.
2. Make the sending task be forever pending on the channel.
3. Let the sending task fill up the channel and then be forever pending (like in Go).
4. Let the sending task know that the channel is disconnected (like in crossbeam-channel).

I don't think there's a right or wrong answer here - it's mostly a personal preference but I do have some *opinions*. :) Different languages and libraries do different things.

If sending into a closed channels succeeds, that means the sending task will keep working forever, which might waste CPU time. I fear this might accidentally get a task into an infinite loop.

**Comment by jjl** [comment]
I've had a think about this, and here's my (possibly wrong :) ) opinion, probably too much inspired by erlang (because I'm building an erlang-style reliability library in rust):

* When a channel is full, backpressure should be applied.
* When a channel is closed sending into it can not possibly be useful, so we should learn that it is shut and take compensatory action.

My suggestion is that send and receive should always return some indication of success or failure (like in crossbeam). I think all of the patterns that make one end responsible for closing the channel force a particular way of dealing with things onto you that's undesirable in a generally useful library. And I always find those APIs quite unintuitive.

Currently, i'm working around this by racing sends with a Timer, but I find this a little bit horrific.

**Comment by jjl** [comment]
Addenda:

* With a method to determine if the channel is closed, I would be able to only race with the timer if it didn't return closed the first time (that is, it becomes a race condition, but one which is is comparatively unlikely so i could better performance in the common case).
* Erlang doesn't have backpressure in the way that these channels do - I'm freestyling a bit here

**Comment by stjepang** [comment]
Returning a result from `send()` makes a lot of sense when we have blocking channels and threads. But async programs have an advantage here. Consider this:

```rust
let (s, r) = piper::chan(10);

Task::spawn(async {
    loop {
        s.send("ping").await;
    }
})
.detach();

for _ in 0..1000 {
    match r.recv().await {
        None => break,
        Some(msg) => println!("{}", msg),
    }
}
drop(r);
```

Here, once we have received 1000 messages and dropped the receiver, the sending task automatically gets canceled, i.e. it is dropped.

How? The reason is that when we drop the receiver, the channel also drops the sending task's `Waker`. At that point, there are no references to the task anymore (remember, a task is kind of like `Arc<dyn Future>`) - so the `async-task` crate behind `smol` decides to cancel the task drop it.

While it may *seem* that the sending task has an infinite loop, in an async context it *does not*. Pretty cool! :)

Clarification: if we didn't detach the task and kept the task reference as in `let task = Task::spawn(async { ... })`, then `task` variable would be the last reference to the task. As soon as *that* variable then gets dropped, the task is canceled.

I think most people don't bother handling the error case in `send()` and decide to just `unwrap()` the result. Also, `s.send(msg).await.unwrap()` is unsightly and annoying to type. In some sense, as I have just shown above, `await` can handle disconnection on its own.

**Comment by jjl** [comment]
Hmm. Indeed, I didn't know about this feature of async-task. Good to hear :)

That said, I wonder how easy it is to break it...

The particular scenario I'm considering at the moment designing [tingle](https://github.com/irrustible/tingle). The current plan for an exit scenario is to be notified on another channel of exit status and then to go on to exit itself by notifying all its linked processes. I suppose I could drop the arbitrary future provided by the user and be sure to drop all my other channels. I guess it would be enough to move them all onto a FuturedUnordered doing a send?

So if that works, I can probably live with it, but I'm definitely finding it confusing to reason about.

**Comment by najamelan** [comment]
> The reason is that when we drop the receiver, the channel also drops the sending task's `Waker`

Looks like fragile coding to me:
1. Doesn't play well with structured concurrency.
2. Executor dependent, eg. not usable in libraries.
3. Often a task will do more than just loop over a send, in which case you don't want it to just drop, so you need to be mindful if this.

The send failing might mean having to restart the receiver task, handling the error, reporting it somewhere, panicking or many other things depending on context.

I do use mechanisms where receiving tasks automatically go away when all senders are dropped, in the sense of "if nobody uses a service anymore, it can be dropped", which seems a little bit less fragile, but it's already quite a problem. Every time you forget to drop them, you get a deadlock (it doesn't slip by unnoticed, but can be annoying to debug). Any time something has to be able to send to itself, it keeps itself alive. Any time two things send to each other, they keep each other alive...

Cancellable tasks help, but then it's nice for the Send to return an error, so you can go and check your code as to why this thing still wants to talk over this channel where you felt it should stop receiving at this point.

**Comment by dignifiedquire** [comment]
some prior discussions on async-std https://github.com/async-rs/async-std/pull/585

**Comment by stjepang** [comment]
I recommend using https://docs.rs/async-channel/
